cachi2/core/package_managers/yarn_classic/main.py

```diff
@@ -203,20 +203,36 @@ def _verify_corepack_yarn_version(source_dir: RootedPath, env: dict[str, str]) -
 
 def _verify_no_offline_mirror_collisions(packages: Iterable[YarnClassicPackage]) -> None:
     """Verify that there are no duplicate tarballs in the offline mirror."""
-    tarballs = []
+    all_tarballs_counter: Counter[str] = Counter()
+    registry_tarballs_counter: Counter[str] = Counter()
+
     for p in packages:
         if isinstance(p, (RegistryPackage, UrlPackage)):
             tarball_name = get_tarball_mirror_name(p.url)
-            tarballs.append(tarball_name)
+            all_tarballs_counter[tarball_name] += 1
+            if isinstance(p, RegistryPackage):
+                registry_tarballs_counter[tarball_name] += 1
+
         elif isinstance(p, GitPackage):
             tarball_name = get_git_tarball_mirror_name(p.url)
-            tarballs.append(tarball_name)
+            all_tarballs_counter[tarball_name] += 1
         else:
             # file, link, and workspace packages are not copied to the offline mirror
             continue
 
-    c = Counter(tarballs)
-    duplicate_tarballs = [f"{name} ({count}x)" for name, count in c.most_common() if count > 1]
+    # check for duplicates, considering cross-type collisions
+    duplicate_tarballs = []
+    for name, count in all_tarballs_counter.items():
+        if count <= 1:
+            continue
+
+        registry_count = registry_tarballs_counter.get(name, 0)
+        non_registry_count = count - registry_count
+
+        # flag if there is at least 1 non-registry package OR a cross-type collision
+        if non_registry_count >= 1:
+            duplicate_tarballs.append(f"{name} ({count}x)")
+
     if len(duplicate_tarballs) > 0:
         raise PackageManagerError(f"Duplicate tarballs detected: {', '.join(duplicate_tarballs)}")
```

Comment:
Although preexisting, as I'm looking at it this isn't a very useful error message because the user may not easily know which package the tarball from the error message belongs to, especially with registry package types.